ui/pages/confirm-transaction/confirm-transaction.component.js

```diff
@@ -19,188 +27,186 @@ import {
+import { isTokenMethodAction } from '../../helpers/utils/transactions.util';
+import { usePrevious } from '../../hooks/usePrevious';
+import {
+  getUnapprovedTransactions,
+  unconfirmedTransactionsListSelector,
+} from '../../selectors';
 import {
   disconnectGasFeeEstimatePoller,
+  getContractMethodData,
   getGasFeeEstimatesAndStartPolling,
   addPollingTokenToAppState,
   removePollingTokenFromAppState,
+  setDefaultHomeActiveTabName,
 } from '../../store/actions';
 import ConfirmTokenTransactionSwitch from './confirm-token-transaction-switch';
 import ConfTx from './conf-tx';
 
-export default class ConfirmTransaction extends Component {
-  static contextTypes = {
-    metricsEvent: PropTypes.func,
-  };
-
-  static propTypes = {
-    history: PropTypes.object.isRequired,
-    totalUnapprovedCount: PropTypes.number.isRequired,
-    sendTo: PropTypes.string,
-    setTransactionToConfirm: PropTypes.func,
-    clearConfirmTransaction: PropTypes.func,
-    mostRecentOverviewPage: PropTypes.string.isRequired,
-    transaction: PropTypes.object,
-    getContractMethodData: PropTypes.func,
-    transactionId: PropTypes.string,
-    paramsTransactionId: PropTypes.string,
-    isTokenMethodAction: PropTypes.bool,
-    setDefaultHomeActiveTabName: PropTypes.func,
-  };
-
-  constructor(props) {
-    super(props);
-    this.state = {};
-  }
-
-  _beforeUnload = () => {
-    this._isMounted = false;
-    if (this.state.pollingToken) {
-      disconnectGasFeeEstimatePoller(this.state.pollingToken);
-      removePollingTokenFromAppState(this.state.pollingToken);
+const ConfirmTransaction = () => {
+  const dispatch = useDispatch();
+  const history = useHistory();
+  const { id: paramsTransactionId } = useParams();
+
+  const [isMounted, setIsMounted] = useState(false);
+  const [pollingToken, setPollingToken] = useState();
+
+  const mostRecentOverviewPage = useSelector(getMostRecentOverviewPage);
+  const sendTo = useSelector(getSendTo);
+  const unapprovedTxs = useSelector(getUnapprovedTransactions);
+  const unconfirmedTransactions = useSelector(
+    unconfirmedTransactionsListSelector,
+  );
+
+  const totalUnapprovedCount = unconfirmedTransactions.length || 0;
+  const transaction = useMemo(() => {
+    return totalUnapprovedCount
+      ? unapprovedTxs[paramsTransactionId] || unconfirmedTransactions[0]
+      : {};
+  }, [
+    paramsTransactionId,
+    totalUnapprovedCount,
+    unapprovedTxs,
+    unconfirmedTransactions,
+  ]);
+
+  const { id, type } = transaction;
+  const transactionId = id && String(id);
+  const isValidERC20TokenMethod = isTokenMethodAction(type);
+
+  const prevParamsTransactionId = usePrevious(paramsTransactionId);
+  const prevTransactionId = usePrevious(transactionId);
+
+  const _beforeUnload = useCallback(() => {
+    setIsMounted(false);
+
+    if (pollingToken) {
+      disconnectGasFeeEstimatePoller(pollingToken);
+      removePollingTokenFromAppState(pollingToken);
     }
-  };
```

Comment:
we can I think use this code in useEffect, may be separate callback is not needed.

Reply:
I think we'll want to keep this since we are passing this callback here:

```
    window.addEventListener('beforeunload', _beforeUnload);
```